example/lib/push_notifications/push_notification_handlers.dart

```diff
@@ -1,20 +1,63 @@
-import 'package:flutter/cupertino.dart';
-
 import 'package:ably_flutter/ably_flutter.dart' as ably;
+import 'package:flutter/cupertino.dart';
 import 'package:flutter/material.dart';
 
 class PushNotificationHandlers {
   static BuildContext? context;
+  static final activationEvents = ably.Push.pushEvents;
+  static final notificationEvents = ably.Push.notificationEvents;
 
   static void setUpEventHandlers() {
-    ably.Push.pushEvents.onUpdateFailed.listen((error) async {
+    activationEvents.onUpdateFailed.listen((error) async {
+      print('onUpdateFailed error:');
       print(error);
     });
-    ably.Push.pushEvents.onActivate.listen((error) async {
+    // Optional
+    activationEvents.onActivate.listen((error) async {
       print(error);
     });
-    ably.Push.pushEvents.onDeactivate.listen((error) async {
+    // Optional
+    activationEvents.onDeactivate.listen((error) async {
       print(error);
     });
   }
+
+  static void setUpMessageHandlers() async {
+    getLaunchMessage();
+
+    notificationEvents.setOnBackgroundMessage(_backgroundMessageHandler);
+
+    notificationEvents.onMessage.listen((remoteMessage) {
+      print('Message was delivered to app while the app was in the foreground: '
+          '$remoteMessage');
+    });
+
+    notificationEvents.setOnShowNotificationInForeground((message) async {
+      print(
+          'Opting to show the notification when the app is in the foreground.');
+      return true;
+    });
+
+    notificationEvents.onNotificationTap.listen((remoteMessage) {
+      print('Notification was tapped: $remoteMessage');
+    });
+  }
+
+  /// You can get the notification which launched the app by a user tapping it.
+  static void getLaunchMessage() {
+    notificationEvents.notificationTapLaunchedAppFromTerminated
+        .then((remoteMessage) {
+      if (remoteMessage != null) {
+        print('The app was launched by the user by tapping the notification');
```

Comment:
I notice that here, and in `_backgroundMessageHandler`, you use multiple `print` statements... I would tend to avoid this approach when they represent the same event as there's a chance they will be separated downstream. Can you not use a single call to `print`, perhaps with a format string or something like that?

Reply:
Thanks. I've refactored this one to print just one line, but also to show a toast when the error happens for the user, since this is in the example app: 32d4d2aa7c6e44ad2afd4ded1c10660c157bee84

Reply:
Nice. 👍